activerecord/test/cases/connection_adapters/connection_handler_test.rb

```diff
@@ -275,6 +275,42 @@ def test_connection_specification_name_should_fallback_to_parent
         ActiveRecord::Base.connection_specification_name = "ActiveRecord::Base"
       end
 
+      unless in_memory_db?
+        def test_connects_to_on_application_record_sets_connection_specification_name_on_active_record_base
+          Object.const_set :ApplicationRecord, ApplicationRecord
+
+          old_config = ActiveRecord::Base.configurations
+          config = { "primary" => { "adapter" => "sqlite3", "database" => "test/db/primary.sqlite3" } }
+          ActiveRecord::Base.configurations = config
+
+          ApplicationRecord.connects_to database: { writing: :primary }
+          assert_equal "ActiveRecord::Base", ApplicationRecord.connection_specification_name
+
+          ActiveRecord::Base.connection_specification_name = "readonly"
+          assert_equal "readonly", ApplicationRecord.connection_specification_name
+        ensure
+          ActiveRecord::Base.configurations = old_config
+          ActiveRecord::Base.establish_connection(:arunit)
+
+          ActiveRecord::Base.connection_specification_name = "ActiveRecord::Base"
+          Object.send :remove_const, :ApplicationRecord
+        end
+      end
+
+      def test_connection_specification_name_can_be_set_on_application_record_directly
+        Object.const_set :ApplicationRecord, ApplicationRecord
+
+        assert_equal "ActiveRecord::Base", MyClass.connection_specification_name
+
+        ApplicationRecord.connection_specification_name = "readonly"
```

Comment:
This is how we've been doing granular connection switching since Rails 5.0 - we call e.g. `ActiveRecord::Base.connection_specification_name = "primary_slow_replica"` to send queries on our primary database to a special replica, without affecting other databases. This worked fine until we added a call to `ApplicationRecord.connects_to` as part of our migration to native multi-db support, which caused only `ActiveRecord::Base` to switch to the slow replica, not any of its subclasses.

This approach does require that you manually establish the connection in the handler by calling `ActiveRecord::Base.connection_handler.establish_connection(:primary_slow_replica)` somewhere, and you're right that it stops working if you switch role, since the connection only exists in the writing hander. Still, it works on Rails 6.0, without establishing a new connection every time like `connected_to(database: :primary_slow_replica)`, and I don't want to have to wait until we're on 6.1 to start introducing `connects_to`.

Reply:
Can I see the call to connects to that broke? We have slow replicas at GitHub, this should work fine:

```ruby
connects_to database: { writing: :primary, reading: :primary_replica, reading_slow: :primary_slow_replica }
```

And then;

```ruby
ActiveRecord::Base.connected_to(role: :reading_slow) do
  # read from slow replica
end
```

Reply:
Using a role doesn't work if you have multiple databases, but only one of them has a slow replica:

```ruby
class ApplicationRecord < ActiveRecord::Base
  connects_to database: { writing: :primary, reading: :primary_replica, reading_slow: :primary_slow_replica
end
class Person < ApplicationRecord; end

class AnimalsRecord < ApplicationRecord
  connects_to database: { writing: :animals, reading: :animals_replica }
end
class Dog < AnimalsRecord; end

ActiveRecord::Base.connected_to(role: :reading_slow) do
  Person.first # queries the primary_slow_replica database
  Dog.first    # raises ActiveRecord::ConnectionNotEstablished
end
```

I could make the code above work by adding `reading_slow: :animals_replica` to `AnimalsRecord`, but now there are two connections to the animals_replica database from every process, which is a significant problem for us.

Now that I've typed that out, I remember that @senny asked about this situation in Basecamp months ago:
https://3.basecamp.com/3076981/buckets/24956/chats/12416418@2824765082

The way we currently achieve this is:

```ruby
ActiveRecord::Base.connection_handler.establish_connection(:primary_slow_replica)

class ActiveRecord::Base
  def with_connection_specification_name(spec_name)
    old_spec_name = connection_specification_name
    self.connection_specification_name = spec_name
    yield
  ensure
    self.connection_specification_name = old_spec_name
  end
end

class ApplicationRecord < ActiveRecord::Base; end
class Person < ApplicationRecord; end

class AnimalsRecord < ApplicationRecord; end
  establish_connection :animals
end
class Dog < AnimalsRecord; end

ActiveRecord::Base.with_connection_specification_name(:primary_slow_replica) do
  Person.first # queries the primary_slow_replica database
  Dog.first    # queries the animals database
end
```

As soon as I add a `connected_to` call to `ApplicationRecord` though, subclasses' connections aren't switched:

```ruby
class ApplicationRecord < ActiveRecord::Base
  connects_to database: { writing: :primary, reading: :primary_replica }
end

ActiveRecord::Base.with_connection_specification_name(:primary_slow_replica) do
  Person.first # queries the primary database
  Dog.first    # queries the animals database
end
```

This patch allows the above approach to work even when `ApplicationRecord.connects_to` has been called.

Reply:
We've been trying to move away from non-class connection_specification_name's because the behavior is confusing, undocumented, and doesn't really make any sense.

I'd be willing to consider merging this as a way to fix your use case but I don't want to support this going forward. Especially because this won't work with https://github.com/rails/rails/pull/40615 and because we changed `ActiveRecord::Base` from `primary` to `ActiveRecord::Base` specifically because it doesn't make sense to support a non-class name `connection_specification_name`.

Reply:
Let's add a deprecation for the setter and move folks away from this hacky way of switching connections. Will the new granular swapping API work for your use case if we remove this functionality entirely in 6.2?

Reply:
Sure, I can add the deprecation.

The new granular swapping API doesn't yet cover my use case, since `ActiveRecord::Base.connected_to` switches all connections, not just the one configured on `ActiveRecord::Base`. There doesn't seem to currently be a way to switch just that one connection.

I think we could make `ApplicationRecord.connected_to` switch just the connection for `ActiveRecord::Base`, in the same way that `ApplicationRecord.connects_to` configures the connection for `ActiveRecord::Base`. I'll open another PR proposing that change.

Reply:
I responded via email so it didn't inline. Calling `ApplicationRecord.connected_to` will only switch connections defined in `ApplicationRecord` and models inheriting from it. `ApplicationRecord`'s behavior exactly matches `AnimalsRecord` in that sense. `ActiveRecord::Base` is global, all other abstract classes are granular.